code/modules/ai/effect_node.dm

```diff
@@ -44,33 +51,40 @@
 		return node
 
 /*
-Current weights are nothing, wait for more ai updates
 Parameter call example
-GetBestAdjNode(list(ENEMY_PRESENCE = -100, DANGER_SCALE = -1)) if these were defined
-This means that the proc will pick out the *best* node
+GetBestAdjNode(list(NODE_LAST_VISITED = -1), IDENTIFIER_XENO)
+Returns an adjacent node that was last visited; when a AI visits a node, it will set NODE_LAST_VISITED to world.time
+
+The score is calculated by what weights are inside of the list/weight_modifiers
+The highest number after multiplying each list/weight by the ones in the above parameter will be the node that's chosen; any nodes that have the same score won't override that node
+
+Generally the number that the weight has before being multiplied by weight modifiers is the "user friendly" edition; NODE_LAST_VISITED represents in deciseconds the time before
+the node has been visited by a particular thing, while something like NODE_ENEMY_COUNT represents the amount of enemies
 */
 
-/obj/effect/ai_node/proc/GetBestAdjNode(list/weight_modifiers)
-	//No weight modifier, return a adjacent random node
-	if(!weight_modifiers)
+/obj/effect/ai_node/proc/GetBestAdjNode(list/weight_modifiers, identifier)
+	//No weight modifiers, return a adjacent random node
+	if(!weight_modifiers || !length(weight_modifiers) || !identifier)
 		return pick(adjacent_nodes)
 
-	var/obj/effect/ai_node/node_to_return = adjacent_nodes[1]
-	var/current_best_node = 0
-	for(var/n in shuffle(adjacent_nodes)) //We keep a score for the nodes and see which one is best
-		var/obj/effect/ai_node/node = n
-		var/current_score = 0
-		for(var/i in 1 to length(weight_modifiers))
-			current_score += GET_WEIGHT(src, i)
+	var/obj/effect/ai_node/node_to_return
+	var/current_best_node_score = -INFINITY
+	var/current_score
+	for(var/obj/effect/ai_node/node in shuffle(adjacent_nodes)) //We keep a score for the nodes and see which one is best
+		current_score = 0
```

Comment:
use
var/i in typed_list
var/atom/cast = i
its faster

Why dont you use shuffle_inplace here we dont need to create a new list do we

Reply:
yeah dont shuffle like this its slow